README.md
tread
=====

A basic terminal RSS feed reader written in Python 3.

![Screenshot](/screenshots/tread.png?raw=true)

Usage
=====

Installation
------------

Install via pip with `pip3 install tread` or download the source and run
`python3 setup.py install`.

Requirements
------------

* Python 3.5
* `curses` (included with Python 3.5 on \*nix systems)
* `sqlalchemy`
* `pyyaml`
* `python-dateutil`
* `requests`
* `beautifulsoup4`
* `imgii`
* `html2text`

Because `curses` is not included on Windows distributions of Python, Windows
users may want to run `tread` in Cygwin (or maybe [bash for Windows](http://www.howtogeek.com/249966/how-to-install-and-use-the-linux-bash-shell-on-windows-10/)
would work). Alternatively, you can make a go of it with one of the
several `curses` implementations available for Windows.

Basic Usage
-----------

Once installed, run `tread`.

Configuration
-------------

By default, `tread` assumes that your configuration file is located at
`~/.tread.yml`. If it isn't, you may specify a configuration file at runtime:
`tread config.yml`.

If the configuration file can't found at runtime, `tread` will create a
configuration file for you with reasonable defaults and copy it to the specified
(or default) location. You may then edit this file to add feeds or change
optional parameters.

### Subscribing to a Feed

While some optional parameters can be configured (who doesn't love tweaking HTTP
request timeout values?), most users will primarily use the configuration
file to specify the feeds to which they want to subscribe. For each feed,
`tread` requires a `name` and a `url`, in the following format:

```yaml
feeds:
  - name: Bad Astronomy
    url: https://www.syfy.com/syfy-wire/topic/bad-astronomy/feed
  - name: Whatever
    url: https://whatever.scalzi.com/feed
  - name: Waypoint
    url: https://waypoint.vice.com/en_us/rss
  - name: Motherboard
    url: https://motherboard.vice.com/en_us/rss
  - name: Pluralistic.net
    url: https://pluralistic.net/feed/
  - name: xkcd
    url: https://xkcd.com/rss.xml
```

Make sure that you get the spacing right (using spaces, not tabs); YAML can be a
little finicky.

### Supported Parsers

Several parsers are available to convert the HTML content found in RSS feeds to
text easily displayed in a terminal. The parser to use is specified using the
configuration file's `parser` field.

Acceptable values for the `parser` field are: `html2text` (default), `lynx`, and
`w3m`.

If you'd prefer to avoid external calls, you probably want to use `html2text`,
which will convert the content to markdown; the `lynx` and `w3m` browsers can
also be used to parse the content (if you have them installed).

Updating Feeds
--------------

Your feeds will be updated periodically while you use `tread`, and you can
manually update a feed by hitting the key dfined in your configuration YAML file
(which defaults to `U`).

If you want to keep your feeds up-to-date even when the program isn't open you
can use `cron` (or something similar) to schedule updates. This is helpful if
you subscribe to a site that posts a lot of content (or has a short RSS history)
and you don't check your feed reader every day.

To update your feeds in a non-interactive mode, simply pass the `--update` (or
`-u`) flag:

```bash
$ tread --update
```

If you want your feeds to be updated daily at 09:00, add the following line to
your `crontab` with `crontab -e`:

```cron
0 9 * * * tread --update
```

On OS X, there are plenty of apps available for scheduling tasks; if you don't
want to install a new application, you can use the builtin `launchd`, although
it can be [a little more complicated](http://alvinalexander.com/mac-os-x/launchd-examples-launchd-plist-file-examples-mac).

Bugs and Feature Requests
=========================

Feature Requests
----------------

* Support for ATOM feeds
* Allow feed URLs to be updated (e.g., maybe with "previous\_url" in YAML?)
* View toggles to display only unread or only starred items (or the combination
  of those two)
* Ability to scroll feed list
* Configurable DB pruning (only keep X days to prevent DB from ballooning)
* Colour support for images
* [bcj](https://github.com/bcj) recommends changing the name to `cuRSSes`

Known Bugs
----------

* When an image is displayed in a feed item and a series resizes occur, way too
  many external requests get generated, and hosts will frequently stop
  responding (and it's slow); solve by caching all ASCII images in a dict by URL
* Resizing the screen results in losing the contents of the messages window (oh
  well)
* Currently ignores the `<sy:updatePeriod>`, `<sy:updateFrequency>`, and
  `<sy:updateBase>` tags in favour of re-fetching each feed at the interval
  specified in the config file

License Information
===================

Written by Gem Newman. [Website](http://spurll.com) | [GitHub](https://github.com/spurll/) | [Twitter](https://twitter.com/spurll)

This work is licensed under Creative Commons [BY-SA 4.0](http://creativecommons.org/licenses/by-sa/4.0/).

Remember: [GitHub is not my CV](https://blog.jcoglan.com/2013/11/15/why-github-is-not-your-cv/).


changes.txt
v0.9, 2016-05-19 -- Initial release
v0.9.1, 2016-05-20 -- Fix attempted import of a nonexistent module
v0.9.2, 2016-05-20 -- Fix crash with feed items that have no title
v0.9.3, 2016-11-23 -- Fix crash with feed items that have no GUID; add manual feed refresh; improve HTML entity parsing
v0.9.4, 2022-07-07 -- Update pyyaml to use safe_load; add logging for yaml load errors
v0.9.5, 2022-07-08 -- Parse feeds as XML instead of HTML, fixing broken item links and open-in-browser functionality
v0.9.6, 2022-07-08 -- Fix width calculations for feed titles containing double-width Unicode characters


main.py
#!/usr/bin/env python3

# Written by Gem Newman. This work is licensed under a Creative Commons
# Attribution-ShareAlike 4.0 International License.


import os
from argparse import ArgumentParser
from curses import wrapper
from functools import partial

from tread.controller import main, update_feeds


def console_main():
    parser = ArgumentParser(description='A simple terminal feed reader.')
    parser.add_argument(
        'config', nargs='?', help='Path to the configuration file to use. '
        'Defaults to ~/.tread.yml.', default='~/.tread.yml'
    )
    parser.add_argument(
        '-u', '--update', help='Instead of running interactively, fetch '
        'updates for all feeds then exit.', action='store_true'
    )
    args = parser.parse_args()

    if args.update:
        update_feeds(os.path.expanduser(args.config))
    else:
        wrapper(
            partial(main, config_file=os.path.expanduser(args.config))
        )


if __name__ == '__main__':
    console_main()


tread/models.py
import curses
from dateutil.parser import parse
from datetime import datetime
from bs4 import BeautifulSoup
from html import unescape
from sqlalchemy import Column, ForeignKey
from sqlalchemy import Integer, Unicode, UnicodeText, DateTime, Boolean
from sqlalchemy.orm import relationship
from sqlalchemy.ext.declarative import declarative_base


Base = declarative_base()


class Feed(Base):
    __tablename__ = 'feeds'

    id = Column(Integer, primary_key=True)
    url = Column(Unicode, index=True)
    name = Column(Unicode)
    main_url = Column(Unicode)
    description = Column(UnicodeText)
    last_refresh = Column(DateTime)

    items = relationship(
        'Item', order_by='desc(Item.date)', back_populates='feed'
    )

    def __init__(self, name, url):
        self.name = name
        self.url = url
        self.main_url = ''
        self.description = ''
        self.last_refresh = None

    @property
    def unread(self):
        return len(list(filter(lambda item: not item.read, self.items)))

    @property
    def starred(self):
        return len(list(filter(lambda item: item.starred, self.items)))

    # Update object from web and write back to DB.
    def refresh(self, db_session, www_session, timeout, log=print):
        log('Refreshing {}...'.format(self.name))
        try:
            r = www_session.get(self.url, timeout=timeout)
        except:
            log('Unable to refresh: no response from {}.'.format(self.url))
            return

        if r.status_code != 200:
            log(
                'Unable to refresh: {} responded with {}.'.format(
                    self.url, r.status_code
                )
            )
            return

        xml = r.text
        soup = BeautifulSoup(xml, 'xml')

        # TODO: Add support for ATOM feeds as well.

        # Nope, just use the name from the config file.
        # self.name = soup.channel.name.string

        self.main_url = soup.channel.link.string
        self.description = soup.channel.description.string

        for item in soup.find_all('item'):
            guid = item.guid.string if item.guid else item.link.string
            row = db_session.query(Item).filter(Item.feed_id == self.id) \
                .filter(Item.guid == guid).scalar()

            if row:
                # Update the item.
                row.title = unescape(
                    item.title.string if item.title.string else ''
                )
                row.url = item.link.string
                row.date = parse(item.pubDate.string)
                row.content = unescape(
                    (item.find('content:encoded') or item.description).string
                )

            else:
                # Create the item.
                row = Item(
                    guid=guid,
                    title=unescape(
                        item.title.string if item.title.string else ''
                    ),
                    url=item.link.string,
                    date=parse(item.pubDate.string),
                    content=unescape(
                        (
                            item.find('content:encoded') or item.description
                        ).string
                    )
                )
                self.items.append(row)

        # Feed has been refreshed.
        self.last_refresh = datetime.utcnow()

        # Write back to DB.
        db_session.add(self)
        db_session.commit()


class Item(Base):
    __tablename__ = 'items'

    id = Column(Integer, primary_key=True)
    guid = Column(Unicode, index=True)
    title = Column(Unicode)
    url = Column(Unicode)
    date = Column(DateTime)
    content = Column(UnicodeText)
    read = Column(Boolean, default=False)
    starred = Column(Boolean, default=False)

    feed_id = Column(Integer, ForeignKey('feeds.id'))
    feed = relationship('Feed', back_populates='items')


class Window:
    border_height = 1
    border_width = 2

    def __init__(
        self, screen, height=None, width=None, row_offset=0, col_offset=0,
        max_lines=1000, border=True, title=''
    ):
        # Size information.
        self.full_height = height if height is not None else curses.LINES
        self.full_width = width if width is not None else curses.COLS
        self.max_lines = max_lines

        # Position information.
        self.row_offset = row_offset
        self.col_offset = col_offset
        self.scroll_pos = 0
        self.next_row = 0

        # Size/offset may be defined relative to full size of terminal.
        if self.full_height < 0:
            self.full_height += curses.LINES
        if self.full_width < 0:
            self.full_width += curses.COLS
        if self.row_offset < 0:
            self.row_offset += curses.LINES
        if self.col_offset < 0:
            self.col_offset += curses.COLS

        # Border and title setup.
        self.border = border
        self.title = title

        # Curses setup.
        self.screen = screen
        self.window = curses.newwin(
            self.full_height, self.full_width, self.row_offset, self.col_offset
        )
        self.pad = curses.newpad(self.max_lines, self.width)

        # Refreshing border also refreshes the title.
        self.refresh_border()

    @property
    def height(self):
        border = Window.border_height if self.border else 0
        return self.full_height - 2 * border

    @property
    def width(self):
        border = Window.border_width if self.border else 0
        return self.full_width - 2 * border

    def write(
        self, string, row_offset=None, col_offset=0, attr=curses.A_NORMAL,
        autoscroll=False, log=None
    ):
        if row_offset is None:
            row_offset = self.next_row

        try:
            self.pad.addstr(row_offset, col_offset, string, attr)
        except curses.error:
            # Lazy way to prevent writing too many buffer lines. Ugh.
            self.max_lines += 100
            self.pad.resize(self.max_lines, self.width)
            self.pad.addstr(row_offset, col_offset, string, attr)

        cursor = self.pad.getyx()
        self.next_row = cursor[0] if cursor[1] == 0 else cursor[0] + 1

        if autoscroll:
            self.scroll(self.next_row - self.height)

    def scroll(self, scroll_pos):
        old_pos = self.scroll_pos
        self.scroll_pos = scroll_pos
        self.constrain_scroll()

        if self.scroll_pos != old_pos:
            self.refresh()

    def scroll_down(self, n=1):
        self.scroll(self.scroll_pos + n)

    def scroll_up(self, n=1):
        self.scroll(self.scroll_pos - n)

    def constrain_scroll(self, first_line=0, last_line=None):
        if last_line is None:
            last_line = self.next_row

        # Prevent negative scroll.
        self.scroll_pos = max(self.scroll_pos, first_line)

        # Prevent scrolling past the content.
        self.scroll_pos = min(self.scroll_pos, last_line - self.height)

    def clear(self):
        self.pad.clear()

    def refresh(self):
        border_height = Window.border_height if self.border else 0
        border_width = Window.border_width if self.border else 0

        try:
            self.pad.refresh(
                self.scroll_pos, 0,
                border_height + self.row_offset,
                border_width + self.col_offset,
                border_height + self.row_offset + self.height - 1,
                border_width + self.col_offset + self.width - 1
            )
        except:
            # Curses is garbage.
            pass

    def refresh_border(self):
        if self.border:
            self.window.border()

        if self.title:
            self.window.addstr(0, self.centre(self.title), self.title)

        self.window.refresh()

    def centre(self, text):
        return max((self.full_width - len(text)) // 2, 0)

    def resize(
        self, new_height, new_width, new_row_offset=None, new_col_offset=None
    ):
        self.full_height = new_height
        self.full_width = new_width

        if new_row_offset is not None:
            self.row_offset = new_row_offset

        if new_col_offset is not None:
            self.col_offset = new_col_offset

        # These should probably be resize (and move) calls, but that's
        # complicated and (apparently) not recommended.
        del self.window
        self.window = curses.newwin(
            self.full_height, self.full_width, self.row_offset, self.col_offset
        )

        del self.pad
        self.pad = curses.newpad(self.max_lines, self.width)

        self.refresh_border()


tread.py
#!/usr/bin/env python

# Written by Gem Newman. This work is licensed under a Creative Commons         
# Attribution-ShareAlike 4.0 International License.                    


import subprocess, requests, yaml, curses, dateutil, textwrap, sys
from shutil import get_terminal_size
from bs4 import BeautifulSoup
from datetime import datetime, timedelta
from dateutil.parser import parse
from imgii import image_to_ascii


with open('config.yml') as f:
    config = yaml.load(f)

# Set up requests to fetch data with retries.
session = requests.Session()
adapter = requests.adapters.HTTPAdapter(max_retries=config.get('retries'))
session.mount('http://', adapter)


class Window:
    border_height = 1
    border_width = 2

    def __init__(
        self, screen,
        height=None, width=None, max_lines=None,
        row_offset=0, col_offset=0,
        border=True, title=''
    ):
        # Size information.
        self.full_height = height if height is not None else curses.LINES
        self.full_width = width if width is not None else curses.COLS
        self.max_lines = config.get('buffer_lines', 1000)   # This is a kludge.
        # TODO: We may need to figure out how to scroll the sidebar.

        # Position information.
        self.row_offset = row_offset
        self.col_offset = col_offset
        self.scroll_pos = 0

        # Size/offset may be defined relative to full size of terminal.
        if self.full_height < 0:
            self.full_height += curses.LINES
        if self.full_width < 0:
            self.full_width += curses.COLS
        if self.row_offset < 0:
            self.row_offset += curses.LINES
        if self.col_offset < 0:
            self.col_offset += curses.COLS

        # Curses setup.
        self.screen = screen
        self.window = curses.newwin(
            self.full_height, self.full_width, self.row_offset, self.col_offset
        )
        self.pad = curses.newpad(self.max_lines, self.width)

        # Border setup.
        if border:
            self.refresh_border()

        # Title setup.
        self.title = title
        if title:
            self.refresh_title()

    @property
    def height(self):
        return self.full_height - 2 * Window.border_height

    @property
    def width(self):
        return self.full_width - 2 * Window.border_width

    def print(self, string, row_offset=0, col_offset=0, attr=curses.A_NORMAL):
        try:
            self.pad.addstr(row_offset, col_offset, string, attr)
        except curses.error:
            # Lazy way to prevent writing too many buffer lines. Ugh.
            pass

    def scroll(self, scroll_pos):
        old_pos = self.scroll_pos
        self.scroll_pos = scroll_pos
        self.constrain_scroll()

        if self.scroll_pos != old_pos:
            self.refresh()

    def scroll_down(self, n=1):
        self.scroll(self.scroll_pos + n)

    def scroll_up(self, n=1):
        self.scroll(self.scroll_pos - n)

    def constrain_scroll(self, last_line=None):
        # Prevent negative scroll.
        self.scroll_pos = max(self.scroll_pos, 0)

        # Prevent scrolling past the content.
        if last_line is not None:
            self.scroll_pos = min(self.scroll_pos, last_line - self.height)

    def clear(self):
        self.pad.clear()

    def refresh(self):
        self.pad.noutrefresh(
            self.scroll_pos, 0,
            Window.border_height + self.row_offset,
            Window.border_width + self.col_offset,
            Window.border_height + self.row_offset + self.height - 1,
            Window.border_width + self.col_offset + self.width - 1
        )
        curses.doupdate()

    def refresh_border(self):
        self.window.border()
        self.window.noutrefresh()
        curses.doupdate()

    def refresh_title(self):
        self.window.addstr(0, centre(self.title, self.full_width), self.title)
        self.window.noutrefresh()
        curses.doupdate()

    def resize(self, height, width):
        # TODO
        pass


class Feed:
    refresh_rate = timedelta(minutes=config.get('refresh', 10))

    def __init__(self, url):
        self.url = url
        self.main_url = ''
        self.title = url
        self.description = ''
        self.items = []

        self.last_refresh = None
        self.refresh(title_only=True)
        # TODO: This is just as slow as loading the whole thing. Use the DB.


    # TODO: Distinguish between load and refresh? When does stuff get loaded from DB?


    # Update object from web and write back to DB.
    def refresh(self, title_only=False, force=False):
        if not force and not self.needs_refresh:
            return

        xml = session.get(self.url, timeout=config.get('timeout')).text
        soup = BeautifulSoup(xml, 'html.parser')

        self.title = soup.channel.title.string

        if title_only:
            return

        self.main_url = soup.channel.link.string
        self.description = soup.channel.description.string
        self.items = [
            Item(
                item.title.string,
                item.link.string,
                parse(item.pubdate.string).astimezone(tz=None),
                item.guid.string,
                (item.find('content:encoded') or item.description).string
            )
            for item in soup.find_all('item')
        ]

        self.last_refresh = datetime.now()

        # TODO: Write back to DB.

    @property
    def needs_refresh(self):
        return (self.last_refresh is None) or (
            datetime.now() - self.last_refresh >= Feed.refresh_rate
        )


class Item:
    browser = config.get('browser', 'lynx')

    def __init__(self, title, url, date, guid, content):
        self.title = title
        self.url = url
        self.date = date
        self.guid = guid
        self.content = content

    def display_content(self, width):
        if Item.browser == 'lynx':
            command = [
                'lynx', '-stdin', '-dump', '-width', str(width), '-image_links'
            ]
            encoding = 'iso-8859-1'
        elif browser == 'w3m':
            command = ['w3m', '-T', 'text/html', '-dump', '-cols', str(width)]
            encoding = 'utf-8'
        else:
            Exception('Unsuported browser: {}'.format(Item.browser))

        output = subprocess.check_output(
            command,
            input=self.content.encode(encoding, 'xmlcharrefreplace'),
            stderr=subprocess.STDOUT
        )
        output = output.decode(encoding, 'xmlcharrefreplace')

        return output


def main(screen):
    # Ensure config['keys'] exists and make all keys uppercase.
    config['keys'] = configure_keys(config.get('keys', dict()))

    # This is the only time the whole screen is ever refreshed. But if you
    # don't refresh it, screen.getkey will clear it, because curses is awful.
    screen.refresh()

    # Turn off visible cursor.
    curses.curs_set(False)

    # Create screen objects.
    # TODO: Ensure screen is large enough to accommodate sidebar and content.
    content, sidebar, menu = init_windows(screen)

    # Add key listing to menu.
    menu.print(menu_text(config['keys'], menu.width))
    menu.refresh()

    # TODO
    # For each feed in config, load contents from the DB.
    # Mark all as "needs refresh".
    # Since the top one is selected, refresh it.
    # Proceed to display.


    # Fetch all feed data.
    feeds = [Feed(url) for url in config['feeds']]

    # Initial selections.
    if len(feeds) > 0:
        feeds[0].refresh()
    selected_feed = 0
    selected_item = 0


    # TODO: Feed selection, item selection, etc. should probably be abstracted
    # into an object as well. These loops are really awkward, as is the line-
    # counting.


    while True:
        current_item = None

        # Print sidebar.
        for i, feed in enumerate(feeds):
            sidebar.print(
                '{:{}}'.format(feed.title, sidebar.width),
                row_offset=i,
                attr=curses.A_REVERSE if i == selected_feed else curses.A_BOLD
            )

            if i == selected_feed:
                current_feed = feed

        # Refresh sidebar.
        sidebar.refresh()

        # Print content.
        line = 0
        for i, item in enumerate(current_feed.items):
            if line >= content.max_lines: break
            # TODO: This is a terrible kludge. At least warn or something.

            content.print(
                '{:{}}{:%Y-%m-%d %H:%M}'.format(
                    item.title, content.width - 16, item.date
                ),
                row_offset=line,
                attr=curses.A_REVERSE if i == selected_item else curses.A_BOLD
            )
            line += 1

            if i == selected_item:
                current_item = item
                line += 1

                # Parse the HTML content.
                parsed_string = item.display_content(content.width)

                # Print it to the screen.
                content.print(parsed_string, row_offset=line)

                line += parsed_string.count('\n') + 1

        # Undo scrolling if content isn't big enough to scroll.
        content.constrain_scroll(line)
        content.refresh()

        # Block, waiting for input.
        key = screen.getkey().upper()

        if key == config['keys']['next_item']:
            if len(current_feed.items) > 0:
                content.clear() # Should be more selective.
                selected_item = (selected_item + 1) % len(current_feed.items)
        elif key == config['keys']['prev_item']:
            if len(current_feed.items) > 0:
                content.clear() # Should be more selective.
                selected_item = (selected_item - 1) % len(current_feed.items)
        elif key == config['keys']['next_feed']:
            content.clear() # Should be more selective.
            sidebar.clear() # Should be more selective.
            selected_feed = (selected_feed + 1) % len(feeds)
            selected_item = 0
            feeds[selected_feed].refresh()
        elif key == config['keys']['prev_feed']:
            content.clear() # Should be more selective.
            sidebar.clear() # Should be more selective.
            selected_feed = (selected_feed - 1) % len(feeds)
            selected_item = 0
            feeds[selected_feed].refresh()
        elif key == config['keys']['scroll_down']:
            content.clear() # Should be more selective.
            content.scroll_down()
        elif key == config['keys']['scroll_up']:
            content.clear() # Should be more selective.
            content.scroll_up()
        elif key == config['keys']['open_in_browser']:
            if current_item is not None:
                open_in_browser(current_item.url)
        elif key == config['keys']['mark_read']:
            pass    # TODO
        elif key == config['keys']['mark_unread']:
            pass    # TODO
        elif key == config['keys']['star']:
            pass    # TODO
        elif key == config['keys']['quit']:
            break

        sidebar.refresh_border()
        sidebar.refresh_title()

    # Write a message function that pops up an overlay window to display a
    # message and pauses for anykey (or enter?)



    # Looks like some unicode characters aren't working:
    # http://xkcd.com/1647/



    # In the DB should store/use guids.
    # Should probably store all of the articles in the DB, and sync them with
    # feeds on launch and on demand.

    # Do something about resize!


    # Warn that this will probably only work in lynx (or implement both...?)
    # Insert output from image conversion on its own lines before the image tag
    # (leave it)
    # use image_to_ascii()

    # Also include html2text as a parser



def init_windows(screen):
    menu_items = 7
    menu_height = menu_items + 2 * Window.border_height

    content = Window(screen, width=-41, col_offset=41, title='TREAD')

    sidebar = Window(
        screen, width=40, height=-menu_height, max_lines=200,
        title='FEEDS'
    )

    menu = Window(
        screen, width=40, height=menu_height, row_offset=-menu_height,
        max_lines=menu_items, title='KEYS'
    )

    return (content, sidebar, menu)


def open_in_browser(url):
    if sys.platform.startswith('linux'):
        subprocess.Popen(['xdg-open', url])
    elif sys.platform.startswith('darwin'):
        subprocess.Popen(['open', url])


def centre(text, width):
    return (width - len(text)) // 2


def configure_keys(existing):
    default = {
        'prev_item': 'K',
        'next_item': 'J',
        'prev_feed': 'H',
        'next_feed': 'L',
        'scroll_up': 'KEY_UP',
        'scroll_down': 'KEY_DOWN',
        'mark_read': 'R',
        'mark_unread': 'U',
        'star': 'S',
        'open_in_browser': 'O',
        'quit': 'Q'
    }

    # Make uppercase.
    existing = {key: value.upper() for key, value in existing.items()}

    # Return default keys, overwriting with the keys that exist in config.
    return {**default, **existing}


def menu_text(keys, width):
    key_width = 17
    value_width = width - key_width
    menu_format = '{:<{}}{:>{}}'

    menu = [
        menu_format.format(
            'Prev/Next Item:', key_width,
            keys['prev_item'] + '/' + keys['next_item'], value_width
        ),
        menu_format.format(
            'Prev/Next Feed:', key_width,
            keys['prev_feed'] + '/' + keys['next_feed'], value_width
        ),
        menu_format.format(
            'Scroll Up/Down:', key_width,
            keys['scroll_up'] + '/' + keys['scroll_down'], value_width
        ),
        menu_format.format(
            'Mark Read/Unread:', key_width,
            keys['mark_read'] + '/' + keys['mark_unread'], value_width
        ),
        menu_format.format(
            'Toggle Star:', key_width, keys['star'], value_width
        ),
        menu_format.format(
            'Open in Browser:', key_width, keys['open_in_browser'], value_width
        ),
        menu_format.format(
            'Quit', key_width, keys['quit'], value_width
        ),
    ]

    return ''.join(menu)


if __name__ == '__main__':
    curses.wrapper(main)
